src/pytest_dbfixtures/port.py

```diff
@@ -41,44 +42,52 @@ def get_port(ports):
     When a range or list of ports is passed `port_for` external package
     is used in order to find a free port.
 
-    :param int|str ports: e.g. 3000, '3000', '3000-3100', '3000,3002', '?'
+    :param int|str|list of tuples and/or sets ports:
+        e.g. 3000, '3000', [(20000, 30000), {48889, 50121}]
     :returns: a random free port
     """
+    if ports == -1:
+        return None
+
     try:
         return int(ports)
-    except ValueError:
+    except TypeError:
         pass
 
     return port_for.select_random(parse_ports(ports))
 
 
 def parse_ports(ports):
     """
-    :param str ports: e.g. '3000', '3000-3100', '3000,3002', '?'
+    :param list of tuples and/or sets:
+        e.g. [(20000, 30000), {48889, 50121}]
     :returns: ports set reflecting specifified ports list/range.
     :rtype set
     """
 
-    if ports == '?':
+    if not ports:
         return None
 
     port_set = set()
 
-    for p in ports.split(','):
-        if '-' not in p:
-            # single, comma-separated port:
-            try:
-                port_set.add(int(p))
-            except ValueError:
-                raise InvalidPortsDefinition(ports)
-        else:
-            # range of ports:
-            try:
-                start, end = p.split('-')
-            except ValueError:
-                raise InvalidPortsDefinition(ports)
+    for element in ports:
+        # tuple, mean range of ports
+        if isinstance(element, tuple):
+            start, end = element
             if end < start:
                 raise InvalidPortsDefinition(ports)
             port_set.update(range(int(start), int(end) + 1))
+        # set, a series of ports
+        elif isinstance(element, set):
+            if all(isinstance(element, int) for e in element):
```

Comment:
this if is not really needed. it seems to be vaste of time, since the else would also cover it, and at the same time . Plus note that for every element in set you check here if set is instance of an integer.